File: lib/Model/user_model.dart

import 'face_features.dart';

class UserModel {
  String? id;
  String? name;
  String? image;
  FaceFeatures? faceFeatures;

  UserModel({
    this.id,
    this.name,
    this.image,
    this.faceFeatures,
  });


  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(

      id: json['id'],
      name: json['name'],
      image: json['image'],
      faceFeatures: FaceFeatures.fromJson(json["faceFeatures"]),
    );
  }


  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'image': image,
      'faceFeatures': faceFeatures?.toJson() ?? {},

    };
  }
}

File: Chat_server/models/user_model.js
const mongoose = require('mongoose');

// تعريف مخطط Points الذي يمثل نقطة ثنائية الأبعاد
const pointsSchema = new mongoose.Schema({
    x: { type: Number, required: true, default: 0 }, // إحداثي X مع قيمة افتراضية
    y: { type: Number, required: true, default: 0 }, // إحداثي Y مع قيمة افتراضية
});

// تعريف مخطط FaceFeatures الذي يمثل ملامح الوجه
const faceFeaturesSchema = new mongoose.Schema({
    rightEar: { type: pointsSchema, required: false },    // الأذن اليمنى
    leftEar: { type: pointsSchema, required: false },     // الأذن اليسرى
    rightEye: { type: pointsSchema, required: false },    // العين اليمنى
    leftEye: { type: pointsSchema, required: false },     // العين اليسرى
    rightCheek: { type: pointsSchema, required: false },  // الخد الأيمن
    leftCheek: { type: pointsSchema, required: false },   // الخد الأيسر
    rightMouth: { type: pointsSchema, required: false },  // زاوية الفم اليمنى
    leftMouth: { type: pointsSchema, required: false },   // زاوية الفم اليسرى
    noseBase: { type: pointsSchema, required: false },    // قاعدة الأنف
    bottomMouth: { type: pointsSchema, required: false }, // الجزء السفلي من الفم
});

// تعريف مخطط UserModel الذي يمثل المستخدم
const userSchema = new mongoose.Schema({
    id: { type: String, required: true }, // معرف المستخدم
    name: { type: String, required: true }, // اسم المستخدم
    image: { type: String, required: true }, // صورة المستخدم
    faceFeatures: { type: faceFeaturesSchema, required: true }, // ملامح الوجه
    registeredOn: { type: Date, default: Date.now }, // تاريخ التسجيل
});

// تحويل المخطط إلى نموذج
const UserModel = mongoose.model('User', userSchema);

module.exports = UserModel;
